File: 3558-number-of-ways-to-assign-edge-weights-i/3558-number-of-ways-to-assign-edge-weights-i.go
func assignEdgeWeights(edges [][]int) int {
    const MOD = 1_000_000_007

	// if a tree has n nodes, it has n - 1 edges. if a tree has m edges, it has m + 1 nodes
	n := len(edges) + 1

	adj := make([][]int, n+1)
	for _, edge := range edges {
		u, v := edge[0], edge[1]
		adj[u] = append(adj[u], v)
		adj[v] = append(adj[v], u)
	}

	maxDepth := 0

	var dfs func(node, parent, depth int)
	dfs = func(node, parent, depth int) {
		if depth > maxDepth {
			maxDepth = depth
		}

		for _, neigh := range adj[node] {
			if neigh != parent {
				dfs(neigh, node, depth+1)
			}
		}
	}

	dfs(1, 0, 0) // start from node 1

    // ans = 2^(k-1) = number of ways with total cost is odd
    ans := 1
    for i := 0; i < maxDepth-1; i++ {
        ans = (ans * 2) % MOD
    }

    return ans
}

// dfs, math, tree
// time: O(n)
// space: O(n)

File: 3558-number-of-ways-to-assign-edge-weights-i/3558-number-of-ways-to-assign-edge-weights-i.py
class Solution:
    def assignEdgeWeights(self, edges: List[List[int]]) -> int:
        MOD = 10**9 + 7

        n = len(edges) + 1
        adj = [[] for _ in range(n + 1)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        max_depth = 0

        def dfs(node: int, parent: int, depth: int) -> None:
            nonlocal max_depth

            if depth > max_depth:
                max_depth = depth
            
            for neigh in adj[node]:
                if neigh != parent:
                    dfs(neigh, node, depth + 1)
            

        dfs(1, 0, 0)

        return pow(2, max_depth - 1, MOD)

# dfs, math, tree
# time: O(n)
# space: O(n)